File: 3_pointer_aliasing/cpp/0/main.cpp
#include <iostream>
#include <memory>
#include <iomanip>
#include <sstream>
#include "main.hpp"

using namespace std;

int main(void) {
    func();
}

void func(void) {
    string *foo = new string("bar");
    string *ptr = foo;
    print(foo, ptr);
    delete foo;
    foo = nullptr;
    print(foo, ptr);
}

void print(string* foo, string* ptr) {
    std::ostringstream s;
    cout << left << setw(10) << "foo";
    cout << left << "ptr" << endl;
    cout << left << setw(10) << foo;
    cout << left << ptr << endl;
    if (foo == nullptr) {
        cout << left << setw(10) << "<invalid>";
    } else {
        s << "\"" << *foo << "\"";
        cout << left << setw(10) << s.str();
    }
    cout << left << "\"" << *ptr << "\"" << "\n" << endl;
}

File: 3_pointer_aliasing/rust/0/src/main.rs
fn main() {
    func_a();
    func_b();
}

fn func_a() {
    let foo = String::from("bar");
    println!("funcA(1): {}", foo);
    let _stolen = foo;
    //println!("funcA(2): {}", foo);
}

fn func_b() {
    let foo = String::from("bar");
    println!("funcB(1): {}", foo);
    take_ownership(foo);
    //println!("funcB(2): {}", foo);
}

fn take_ownership<T>(_: T) { }
